### ISSUE TITLE
When fine-tuning the MMS model, I encountered NaN loss. How can I solve this issue?

### ISSUE THREAD

**Issue by yuhangear** [facebookresearch/fairseq#5245]
### Before asking:
1. search the issues.
2. search the docs.

#### What is your question?
I am trying to train a bilingual model for English and Indonesian. During the training process, the loss initially decreases, but after more steps, the training loss suddenly increases and the predicted logits are all NaN values.
I am using an adapter for English to initialize the adapter of the bilingual model.
In the early stages of training, the loss can decrease and some correct results can be predicted.
Below is the change in loss.
![image](https://github.com/facebookresearch/fairseq/assets/31382694/4b16e4bc-2cdf-416e-80b7-f2e9353c6346)

Below is the logits obtained in the forward function and decoded.
![image](https://github.com/facebookresearch/fairseq/assets/31382694/8bd75970-4c27-492e-831a-cafe98372cbb)

#### Code
This is my config.

model = Wav2Vec2ForCTC.from_pretrained(
    "mms-1b-all-fine-tune/checkpoint-1600",
    attention_dropout=0.0,
    hidden_dropout=0.0,
    feat_proj_dropout=0.0,
    layerdrop=0.0,
    pad_token_id=processor.tokenizer.pad_token_id,
    vocab_size=len(processor.tokenizer),
    ignore_mismatched_sizes=True,
    target_lang=target_lang,
)

model.freeze_base_model()

adapter_weights = model._get_adapters()
for param in adapter_weights.values():
    param.requires_grad = True

from transformers import TrainingArguments

training_args = TrainingArguments(
  output_dir="./mms-1b-all-fine-tune2",
  per_device_train_batch_size=4,
  evaluation_strategy="steps",
  num_train_epochs=4,
  gradient_checkpointing=True,
  fp16=False,
  save_steps=400,
  eval_steps=200,
  logging_steps=200,
  learning_rate=1e-3,
  warmup_steps=100,
  save_total_limit=2,
  push_to_hub=False,
  max_grad_norm=0.5

)

training_args.set_training(batch_size=4,max_steps=20000,gradient_accumulation_steps=4,learning_rate=1e-3)

from transformers import Trainer

trainer = Trainer(
    args=training_args,
    model=model,
    train_dataset=train_dataset,
    eval_dataset=dev_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
    tokenizer=processor.feature_extractor,
)

processor.save_pretrained(training_args.output_dir)

trainer.train()

#### What have you tried?
I have tried the following attempts, but the loss eventually becomes NaN.

Try reducing the learning rate to 1e-6
1.Consider initializing the adapter randowly.
2.Use only Indonesian for training, or train with both English and Indonesian.
3.Consider using dropout, 
4.Reduce gradient accumulation.
5.Set fp16 to False.

**Comment by yuhangear** [comment]
I used a brute force method, which is to check if there are NANs in the gradients during training and set the gradients to zero if there are. So far, there have been no NANs during training, but I'm not sure how this method will affect the performance of the model. I hope experts can explain what negative impacts this approach might have.


    def check_gradients(self,model):
        for param in model.parameters():
            if param.grad is None:
                continue
            if torch.isnan(param.grad).any():
                print("NaN detected in gradients!")
                model.zero_grad()
                print(" NaN detected  and set grad zero")
                return

**Comment by qunash** [comment]
Hi, I'm also interested in fine-tuning the MMS model. Were you able to succesfully fine-tune it?

**Comment by vineelpratap** [comment]
Hi, can you follow this tutorial for finetuning - https://huggingface.co/blog/mms_adapters
